### PR TITLE
Add tests to the hasher

### PR DESCRIPTION
**Pull request by npmccallum**
First we made the hasher produce the same output as the Linux kernel SGX
selftest. Then we made a bunch of tests to verify the algorithm doesn't
accidentally change. These tests have not yet been tested against the Intel
CPU, so they might still be incorrect in some corner cases. However, the
algorithm should be fairly correct at this point.

### CODE REVIEW

File: sgx-hasher/src/lib.rs [hunk 1]
```diff
@@ -38,46 +36,185 @@ impl Hasher {
...
+
+        let contents = sig::Contents::new(
+            misc::MiscSelect::default(),
+            attr::Attributes::default(),
+            [0u8; 32],
+            0,
+            0,
+        );
+
+        let secs = secs::Secs::new(0, size.next_power_of_two(), 1, &contents);
```
Comment (lkatalin) [inline sgx-hasher/src/lib.rs]: So the `Secs` has a base of `0`, a size that is larger than the input size by a power of two, an `ssa` size of `1`, and the default `contents` from above. It may be helpful to include this as a comment.

File: sgx-types/src/page.rs [hunk 2]
```diff
@@ -67,8 +65,20 @@ impl AsRef<[u8]> for SecInfo {
...
-        Self { flags, class }
+    pub fn reg(flags: Flags) -> Self {
+        Self {
+            flags,
+            class: Class::Reg,
+            reserved: Padding::default(),
+        }
+    }
+
+    pub fn tcs() -> Self {
```
Comment (lkatalin) [inline sgx-types/src/page.rs]: This makes sense.

File: sgx-hasher/src/lib.rs [hunk 3]
```diff
@@ -38,46 +36,185 @@ impl Hasher {
...
+    //
+    // The canonical source of correctness for this algorithm is, of
+    // course, the Intel SGX CPU/microcode/ME. If you can demonstrate a
+    // a case where we don't match this, we will happily change our ANSWERs.
+
+    const DATA: [u8; 4096] = [123u8; 4096];
+    const LONG: [u8; 8192] = [123u8; 8192];
+
+    fn hash(input: &[(&[u8], SecInfo)]) -> [u8; 32] {
+        let size = input.iter().fold(0, |c, x| c + x.0.len() as u64);
```
[the remark this answers is not in the document]
Comment (npmccallum) [reply sgx-hasher/src/lib.rs]: I added a comment in the latest push.

File: sgx-hasher/src/lib.rs (hunk 1 above)
Comment (npmccallum) [reply sgx-hasher/src/lib.rs]: I added a comment in the latest push.

File: sgx-types/src/sig.rs [hunk 4]
```diff
@@ -1,6 +1,22 @@
 use super::{attr::Attributes, misc::MiscSelect, utils::Padding};
 
 #[derive(Copy, Clone)]
+struct Number([u8; 384]);
```
Comment (lkatalin) [inline sgx-types/src/sig.rs]: This is a good idea since we're using this type of array multiple times, but could use a more descriptive name. What sort of Number is it? Why is a Number 384 bytes? Etc.

File: sgx-sys/src/lib.rs [hunk 5]
```diff
@@ -76,6 +76,8 @@ mod test {
     fn builder_new() {
         let contents = sig::Contents::new(
```
Comment (lkatalin) [inline sgx-sys/src/lib.rs]: This could possibly use a comment to say the second `misc` and `attr` are the masks.

File: sgx-types/src/sig.rs (hunk 4 above)
Comment (npmccallum) [reply sgx-types/src/sig.rs]: Fixed.

File: sgx-sys/src/lib.rs (hunk 5 above)
Comment (npmccallum) [reply sgx-sys/src/lib.rs]: I actually haven't fully understood this part of SGX yet. I didn't take the masks in the constructor before because I was being lazy. But the `selftest` code required me to differentiate between the two.
